File: 程序员面试金典/4.12求和路径/pathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int pathSum(TreeNode root, int sum) {
        if (root == null)
            return 0;
        return search(root, sum) + pathSum(root.left, sum) + pathSum(root.right, sum);
    }

    private int search(TreeNode root, int sum) {
        if (root == null)
            return 0;
        sum -= root.val;
        return (sum == 0 ? 1 : 0) + search(root.left, sum) + search(root.right, sum);
    }

}

File: 程序员面试金典/4.12求和路径/pathSum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        if not root:
            return 0
        return self.search(root, sum) + self.pathSum(root.left, sum) + self.pathSum(root.right, sum)

    def search(self, root, s):
        if not root:
            return 0
        s -= root.val
        return int(s == 0) + self.search(root.left, s) + self.search(root.right, s)
